NIFI-4289 - InfluxDB put processor

mans2singh opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

mans2singh commented:
Hey Nifi Team:  Please let me know your feedback/comments on this processor.  Thanks.

mans2singh commented:
Hey Folks - Please let me know your feedback/suggestions for this time-series db processor.  Thanks

joewitt commented on nifi-assembly/NOTICE:
```diff
@@ -1543,6 +1543,7 @@ The following binary components are provided under the MIT License.  See project
 
   (MIT License) EWS Java API (com.microsoft.ews-java-api:ews-java-api:2.0 - https://github.com/OfficeDev/ews-java-api)
   (MIT License) libffi (libffi-3.2.1 - http://sourceware.org/libffi/)
+  (MIT License) InfluxDB (org.influxdb:influxdb-java:2.7 https://github.com/influxdata/influxdb-java/)  
```
there should not be any (MIT License) section in our NOTICE.  These all belong in the LICENSE.  I will move these other ones in the current master though

joewitt commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-nar/src/main/resources/META-INF/NOTICE:
```diff
@@ -0,0 +1,28 @@
+
+The following binary components are provided under the MIT License
+
+  The MIT License (MIT)
```
same comment as before.  No MIT license section belongs in NOTICE.  Only LICENSE.  thanks

mans2singh commented:
@joewitt

I've removed the MIT License statements.  Please let me know if there are any other comments/recommndations.

Thanks

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-nar/pom.xml:
```diff
@@ -0,0 +1,44 @@
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-influxdb-bundle</artifactId>
+        <version>1.4.0-SNAPSHOT</version>
```
1.6.0-SNAPSHOT

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/pom.xml:
```diff
@@ -0,0 +1,66 @@
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-influxdb-bundle</artifactId>
+        <version>1.4.0-SNAPSHOT</version>
```
1.6.0-SNAPSHOT

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/pom.xml:
```diff
@@ -0,0 +1,43 @@
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-nar-bundles</artifactId>
+        <version>1.4.0-SNAPSHOT</version>
```
1.6.0-SNAPSHOT

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/pom.xml:
```diff
@@ -0,0 +1,43 @@
+            <dependency>
+                <groupId>org.apache.nifi</groupId>
+                <artifactId>nifi-influxdb-processors</artifactId>
+                <version>1.4.0-SNAPSHOT</version>
```
1.6.0-SNAPSHOT

MikeThomsen commented on pom.xml:
```diff
@@ -1137,6 +1142,12 @@
+            <dependency>
+                <groupId>org.apache.nifi</groupId>
+                <artifactId>nifi-influxdb-nar</artifactId>
+                <version>1.4.0-SNAPSHOT</version>
```
1.6.0-SNAPSHOT

MikeThomsen commented:
@mans2singh Do you have an example of valid JSON input I can use to test this? I grabbed an example from here: https://github.com/influxdata/influxdb-python

That's the Python client. I'm using this in a GenerateFlowFile processor:

``
    {
        "measurement": "cpu_load_short",
        "tags": {
            "host": "server01",
            "region": "us-west"
        },
        "time": "${now():toString()}",
        "fields": {
            "value": 0.64
        }
    }
``

It doesn't like that. Throws this exception:

```
org.influxdb.InfluxDBException: {"error":"unable to parse '{': missing fields\nunable to parse '\"measurement\": \"cpu_load_short\",': invalid field format\nunable to parse '\"tags\": {': invalid field format\nunable to parse '\"host\": \"server01\",': invalid field format\nunable to parse '\"region\": \"us-west\"': invalid field format\nunable to parse '},': missing tag key\nunable to parse '\"time\": \"Sun Jan 21 16:48:12 EST 2018\",': invalid field format\nunable to parse '\"fields\": {': invalid field format\nunable to parse '\"value\": 0.64': invalid field format\nunable to parse '}': missing fields\nunable to parse '}': missing fields"}

	at org.influxdb.impl.InfluxDBImpl.execute(InfluxDBImpl.java:511)
	at org.influxdb.impl.InfluxDBImpl.write(InfluxDBImpl.java:325)
	at org.apache.nifi.processors.influxdb.PutInfluxDB.writeToInfluxDB(PutInfluxDB.java:169)
	at org.apache.nifi.processors.influxdb.PutInfluxDB.onTrigger(PutInfluxDB.java:149)
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1122)
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:147)
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47)
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:128)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

If you can help me out with the test data, I can keep going with the review.

mans2singh commented:
@MikeThomsen - 

Thanks for your review comments.

I've updated the version and rebased the code.  

Regarding the error - InfluxDB uses line protocol for writing data (I've mentioned that in the capabilities section of the processor) - [https://docs.influxdata.com/influxdb/v1.4/write_protocols/line_protocol_tutorial/] to for writing data.  

I have integration/unit test that use the protocol for good and bad format scenarios -

- integration test - [https://github.com/apache/nifi/pull/2101/files#diff-6aa4a2accea7ae7b489e2a219fb777f2]
- unit tests - [https://github.com/apache/nifi/pull/2101/files#diff-9f4273a51b147a9e22085ef842726664]

If you are using the flow file generator with the line protocol - the timestamp (last field) is a number - so please use `${now():toNumber()}` rather than `${now():toString()}` as shown in the sample measurement below:
`water,country=US,city=sf rain=1,humidity=0.6 ${now():toNumber()}`

Please let me know of you have any additional comments/recommendations or if there is anything else required.

Thanks

MikeThomsen commented:
I'm going to try to get some time today and tomorrow to go through the code itself, but I was able to build a flow that uses the line protocol as described, and it successfully inserted multiple entries into InfluxDB. So... so far, so good.

mans2singh commented:
@MikeThomsen - Thanks for your feedback.  Please let me know if I can assist in anyway.  Mans

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+        try {
+            influxDB = makeConnection();
+        } catch(Exception e) {
+            getLogger().error("Error while getting connection " + e.getLocalizedMessage(),e);
```
You should use the same "{}" syntax here.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+            getLogger().error("Error while getting connection " + e.getLocalizedMessage(),e);
+            throw new RuntimeException("Error while getting connection" + e.getLocalizedMessage(),e);
+        }
+        getLogger().info("InfluxDB connection created for host",
```
I think you're missing a "{}" here for string substitution.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    protected InfluxDB makeConnection() {
+        if ( StringUtils.isBlank(username) || StringUtils.isBlank(password) )
```
Please wrap in curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+        if ( StringUtils.isBlank(username) || StringUtils.isBlank(password) )
+            return InfluxDBFactory.connect(influxDbUrl);
+        else
+            return InfluxDBFactory.connect(influxDbUrl, username, password);
```
Curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+
+    @OnStopped
+    public void close() {
+        getLogger().info("Closing connection");
```
Not sure we really need this. Maybe something like this:
```
if (getLogger().isDebugEnabled()) {
    getLogger().debug(....)
}
```

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+    @OnStopped
+    public void close() {
+        getLogger().info("Closing connection");
+        if ( influxDB != null )
```
Missing curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/PutInfluxDB.java:
```diff
@@ -0,0 +1,176 @@
+            .required(true)
+            .defaultValue(CONSISTENCY_LEVEL_ONE.getValue())
+            .expressionLanguageSupported(true)
+            .allowableValues(CONSISTENCY_LEVEL_ONE,CONSISTENCY_LEVEL_ANY,CONSISTENCY_LEVEL_ALL,CONSISTENCY_LEVEL_QUORUM)
```
Nit: unneeded white space.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITPutInfluxDBTest.java:
```diff
@@ -0,0 +1,189 @@
+ * on local host with default port and has database test with table test. Please set user
+ * and password if applicable before running the integration tests.
+ */
+@Ignore("Comment this out for running tests against a real instance of InfluxDB")
```
Maven has the ability to do integration tests. You might want to think about setting up a new profile to do that instead of using Ignore. I have [an example here on this PR](https://github.com/MikeThomsen/nifi/commit/cbc2d61f2d98fdf29ff6f1a46fbb524691fd11c0#diff-9fbcd1bfda73f61e6c31fb2fcb3371f3R111).

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITPutInfluxDBTest.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    protected void checkError(QueryResult result) {
+        if ( result.hasError() )
```
Please add curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITPutInfluxDBTest.java:
```diff
@@ -0,0 +1,189 @@
+    @After
+    public void tearDown() throws Exception {
+        runner = null;
+        if ( influxDB != null )
```
Curly brackets...

MikeThomsen commented:
Everything I insert gets defaulted to some time around the epoch. I tried your sample, but it doesn't seem to work for me at least (data gets written, but it doesn't have the appropriate timestamp):

```
water,country=US,city=sf rain=1,humidity=0.6 ${now():toNumber()}
```

Ex output:

```
water,country=US,city=sf rain=1,humidity=0.6 1516999446718
```

Using this query in Chronograf, I get the following results:

```
SELECT * FROM "test_data"."autogen"."water"
```

<img width="1032" alt="screen shot 2018-01-26 at 3 40 27 pm" src="https://user-images.githubusercontent.com/108184/35459708-4ccb9ba6-02af-11e8-852f-7aeabd1806dc.png">

So overall LGTM on code quality, but we need to figure out if this is just something wrong on my end or with the processor before someone merges it.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,142 @@
+    public void onScheduled(final ProcessContext context) {
+        username = context.getProperty(USERNAME).getValue();
+        password = context.getProperty(PASSWORD).getValue();
+        influxDbUrl = context.getProperty(INFLUX_DB_URL).getValue();
```
I think it would be a better practice to not have these values stored as fields on the class. They get reset every time the user stops and restarts the processor, so I don't think we need to keep them around after the connection is established.

MikeThomsen commented:
@mans2singh Figured out the issue, it's a time conversion. I added ```:multiply(1000000)``` to the EL and it worked. InfluxDB uses nanoseconds; I was supplying milliseconds.

@joewitt FWIW +1 LGTM on merging.

mans2singh commented:
Hello @MikeThomsen 

Thanks for your feedback. Based on your feedback I've update the code (added braces, corrected logging, removed state variables not required from the put processor and added integration tests profile).

Regarding the timestamp issue - I forgot to mention that InfluxDB requires timestamp in nano-seconds resolution so we need to multiply the timestamp with million (I've updated the documentation also).  I've tested with the following template in generate processor:

`water,country=US,city=sf rain=1,humidity=0.6 ${now():toNumber():multiply(1000000) }`

and I see the following in the influx shell:

```
time                     city country humidity rain
----                     ---- ------- -------- ----
2018-01-27T23:31:46.331Z sf   US      0.6      1
```
Please let me know if you have any additional advice/comments for me.

Thanks again.

Mans

MikeThomsen commented:
@mattyb149 Do you have a recommendation for the validator [to use here]( https://github.com/apache/nifi/pull/2101#issuecomment-361025124)?

mattyb149 commented:
Probably NON_EMPTY_VALIDATOR?

mans2singh commented:
@mattyb149 @MikeThomsen 

If I use `NON_EMPTY_VALIDATOR` - I get the validation exception. I am including the exception, property definition and unit test snippets below.  Please let me know if I am missing anything.

Thanks.

Here is the unit test validation exception with `NON_EMPTY_VALIDATOR`:

`
java.lang.AssertionError: Processor has 1 validation failures:
'influxdb-username' validated against '' is invalid because influxdb-username cannot be empty
`

Here is how I am defining the property

```
    public static final PropertyDescriptor USERNAME = new PropertyDescriptor.Builder()
            .name("influxdb-username")
            .displayName("Username")
            .required(false)
            .description("Username for accessing InfluxDB")
            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
            .build();
```
and here is the unit test:

```
    @Test
    public void testEmptyUsername() {
        runner.setProperty(PutInfluxDB.USERNAME,"");
        runner.assertValid();
    }
```

MikeThomsen commented:
@mattyb149 That sounds like the validator is being called even when it's not a required property. Is there a good reason for NiFi to do that?

mattyb149 commented:
By setting the NON_BLANK or NON_EMPTY validator, you are prohibiting someone from using an empty string as a username or password, which is what you want, right? So that unit test should fail because you are setting an empty string on USERNAME and then asserting that it is valid, when it is not.

MikeThomsen commented:
@mans2singh This works in the unit test per @mattyb149 's point:

```
     runner.assertValid();
     runner.setProperty(PutInfluxDB.PASSWORD, "");
     runner.assertNotValid();
```

And that's with `NON_EMPTY_VALIDATOR` being used.

mans2singh commented:
@MikeThomsen @mattyb149 

I've changed the validator and updated the unit test cases for empty username and password.

Please let me know if you have any other comments/recommendations.

Thanks

Mans

MikeThomsen commented:
Testing it out now. Will let you know if actually using it I observe anything weird.

MikeThomsen commented:
+1 LGTM.

MikeThomsen commented:
Here's a simple test that generates a new write every 3 seconds.

[PutInfluxDb_Test.xml.gz](https://github.com/apache/nifi/files/1671417/PutInfluxDb_Test.xml.gz)

This is the query I am using in Chronograf:

```
SELECT mean("humidity") AS "mean_humidity" FROM "test_data"."autogen"."water" WHERE time > now() - 1h GROUP BY :interval: FILL(null)
```

And here is a docker-compose file you can use to set up the backend:

```
version: '2'

services:
  influxdb:
    image: influxdb:latest
    #volumes:
    #  - ./data/influxdb:/var/lib/influxdb
    ports:
      - "8086:8086"
      - "8083:8083"
      - "2003:2003"
    environment:
      INFLUXDB_GRAPHITE_ENABLED: "true"
      INFLUXDB_ADMIN_ENABLED: "true"
  # Define a Chronograf service
  chronograf:
    image: chronograf
    #volumes:
    #  - ./data/chronograf:/var/lib/chronograf
    ports:
      - "10000:10000"
      - "8888:8888"
    links:
      - influxdb
```

Web UI on port 8888 for testing.

mans2singh commented:
@MikeThomsen 

Thanks for the test file and docker-compose.  Let me know if there is anything else required.

MikeThomsen commented:
@mans2singh LGTM, but we need a committer to sign off now.

mattyb149 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,143 @@
+
+    public static final PropertyDescriptor INFLUX_DB_URL = new PropertyDescriptor.Builder()
+            .name("influxdb-url")
+            .displayName("InfluxDB connection url")
```
Consider supporting Expression Language here. Although it might not make sense to use FlowFile attributes (since it would be less efficient to have a connection per execution), supporting EL would allow the use of the Variable Registry, Process Group variables, etc. to set the URL, which helps when promoting flows from dev -> test -> production for example.

mattyb149 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,143 @@
+    public void onScheduled(final ProcessContext context) {
+        String username = context.getProperty(USERNAME).getValue();
+        String password = context.getProperty(PASSWORD).getValue();
+        String influxDbUrl = context.getProperty(INFLUX_DB_URL).getValue();
```
If you do add EL support to fields, remember to evaluate the expressions (you won't need to pass in a parameter) here.

mattyb149 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,143 @@
+        maxRecordsSize = context.getProperty(MAX_RECORDS_SIZE).asDataSize(DataUnit.B).longValue();
+
+        try {
+            influxDB = makeConnection(username, password, influxDbUrl);
```
It is not best practice to make an external connection (especially one with long or infinite timeout) in the OnScheduled method; rather it is recommended to have synchronized code in onTrigger that will initialize the connection only once if it has not been successfully initialized before. Stopping the processor usually resets the initialization "flag" such that starting it again will cause re-init on the first onTrigger() to be invoked.

mattyb149 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/AbstractInfluxDBProcessor.java:
```diff
@@ -0,0 +1,143 @@
+
+    protected InfluxDB makeConnection(String username, String password, String influxDbUrl) {
+        if ( StringUtils.isBlank(username) || StringUtils.isBlank(password) ) {
+            return InfluxDBFactory.connect(influxDbUrl);
```
In order to set a connection timeout on the underlying OkHttp client (see my other comments), you'll need to create that client yourself and pass it into the [InfluxDbFactory.connect](https://github.com/influxdata/influxdb-java/blob/influxdb-java-2.8/src/main/java/org/influxdb/InfluxDBFactory.java#L59) method

mattyb149 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/PutInfluxDB.java:
```diff
@@ -0,0 +1,176 @@
+@EventDriven
+@SupportsBatching
+@Tags({"influxdb", "measurement","insert", "write", "put", "timeseries"})
+@CapabilityDescription("Processor to write the content of a FlowFile (in line protocol https://docs.influxdata.com/influxdb/v1.3/write_protocols/line_protocol_tutorial/) to InfluxDB (https://www.influxdb.com/). "
```
The URL can/will become a dead link in time, perhaps refer to the Line Protocol and say "consult InfluxDB documentation for more details" or something like that.

mattyb149 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/PutInfluxDB.java:
```diff
@@ -0,0 +1,176 @@
+            return;
+        }
+
+        if ( flowFile.getSize() > maxRecordsSize) {
```
This seems a bit stern, what would a user do if the size of the records is too large? Can it be split using NiFi processors? If not, it seems like the data would have no place to go.

MikeThomsen replied:
@mans2singh Correct me if I am wrong, but I think I saw something in the documentation that said that InfluxDB's REST handler only accepts a few MB of data/request (5MB/request?). Is that right?

mans2singh replied:
I could not find any documentation in the InfluxDB docs.

mans2singh commented:
@mattyb149 

I've updated the code based on your review feedback (added MIT license,  updated property description, added expression language support for db name and size properties, and added connection timeout property, updated documentation regarding line protocol). I've also added retry relation (for handling socket timeout), and max size exceed relation to handle if records size exceeds limit.  Unit/integration tests are also updated for these changes.

Regarding max size limit - this is limit on the size of records being posted to InfluxDB.  The processor can push a batch of records separated by new line to InfluxDB. I wanted to give the admin some control over how much data is being posted to InfluxDB.  If the size is exceeded, the size exceed relation can be used to handle this scenario.

Please let me know if you have any additional recommendations.

Thanks again for your thoughtful feedback.

mans2singh commented:
@mattyb149 @MikeThomsen 

Please let me know if you have any addiional comments/recommendations for this processor.

Thanks

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/PutInfluxDB.java:
```diff
@@ -78,18 +81,33 @@
+    static final Relationship REL_RETRY = new Relationship.Builder().name("retry")
+            .description("FlowFiles were not saved to InfluxDB due to retryable exception are routed to this relationship").build();
+
+    static final Relationship REL_MAX_SIZE_EXCEEDED = new Relationship.Builder().name("failure-max-size")
```
Something to consider here...

With PutHBaseRecord and (still pending merge) DeleteHBaseRow, I had a similar situation. Users could easily chuck several hundred thousand HBase operations at the processor all at once. So what was suggested to me, and I did with both of them, was to break up the incoming flowfile into chunks and then add a "retry.index" attribute to the flowfile if it failed. That way, users could loop REL_RETRY to the processor and get everything ingested.

Though that might not apply in this case because InfluxDB doesn't have an ID field that I know of. If you replay an event with the same timestamp, does it overwrite or does it just add a new one?

mans2singh commented:
@MikeThomsen - 

If the fields, tags and timestamp for a measurement are the same, they are considered to be the same record. 

Regarding size limit - I did not find any mention of size limit for posting data in the influxdb docs. I think this all depends on use cases and with the size limit processor property available, the nifi admin configure the values based on their influx db cluster and load.

Let me know if I have missed anything or anything else required.

Thanks again for your advice/comments.

MikeThomsen commented:
@pvillard31 @alopresto Could one of you take a look? I think @mans2singh has done an excellent job addressing our concerns.

pvillard31 commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/pom.xml:
```diff
@@ -0,0 +1,88 @@
+                        <artifactId>maven-surefire-plugin</artifactId>
+                        <configuration>
+                            <excludes>
+                                <exclude>**/IT*.java</exclude>
```
I'm not sure this is really needed as we're already defining what tests have to be included in the root pom XML file.
````xml
<includes>
    <include>**/*Test.class</include>
    <include>**/Test*.class</include>
    <include>**/*Spec.class</include>
</includes>
````

mans2singh replied:
@pvillard31 

If I remove this profile/configuration from the pom.xml and run `mvn clean install` in that directory, the integration tests get executed and fail if InfluxDb is not running on the local server.  

```
Tests in error: 
  ITPutInfluxDBTest.setUp:58 » InfluxDBIO java.net.ConnectException: Failed to c...
  ITPutInfluxDBTest.setUp:58 » InfluxDBIO java.net.ConnectException: Failed to c...
  ITPutInfluxDBTest.setUp:58 » InfluxDBIO java.net.ConnectException: Failed to c...
  ITPutInfluxDBTest.setUp:58 » InfluxDBIO java.net.ConnectException: Failed to c...
  ITPutInfluxDBTest.setUp:58 » InfluxDBIO java.net.ConnectException: Failed to c...
  ITPutInfluxDBTest.setUp:58 » InfluxDBIO java.net.ConnectException: Failed to c...
```

Can you please let me know how I can skip running the integration tests without using this profile ?

Thanks for @MikeThomsen and your feedback.

pvillard31 replied:
Oh my bad, I didn't pay close attention... Usually the IT classes do not include the word Test. That's why I thought the root profile would be enough. I'd suggest to rename the IT classes by removing the Test word to be consistent with the other bundles.

mans2singh commented:
@pvillard31 - 

You were right - After renaming the integration tests they don't get executed by default mvn install.  Also removed test profile from pom.xml as recommended.

Please let me know if you have any additional recommendation.

Thanks again.

pvillard31 commented:
It looks OK to me, I'd just suggest allowing EL on username/password. Will probably be useful with the Registry stuff. I know sensitive properties are not handled by the registry work yet, but that could be in the future.

Before giving a final +1, I'd appreciate another opinion on L&N aspects but it LGTM. Thanks for all your work @mans2singh

mattyb149 commented:
The L&N LGTM :) the MIT license reference goes in the assembly LICENSE and the NAR license (which it is), and nothing in the NOTICE unless its transitive dependencies have NOTICEs, I verified only Apache Commons Lang does, and it is included.

mans2singh commented:
@pvillard31 @mattyb149 @MikeThomsen 

I've added expression language support for username and password.

Please let me know if there is any other recommendation.

Thanks

Mans

pvillard31 commented:
Did some additional local tests. LGTM.
+1, merging to master, thank you all.

mans2singh commented:
Thanks @MikeThomsen @pvillard31 @mattyb149 @joewitt @jskora and everyone for your advice/support.
